lib/stdlib/src/erl_tar.erl

```diff
@@ -1253,37 +1259,38 @@ parse_numeric(<<First, _/binary>> = Bin) ->
 
 parse_octal(Bin) when is_binary(Bin) ->
     %% skip leading/trailing zero bytes and spaces
-    do_parse_octal(Bin, <<>>).
-do_parse_octal(<<>>, <<>>) ->
-    0;
+    do_parse_octal(Bin, 0).
+
 do_parse_octal(<<>>, Acc) ->
-    case io_lib:fread("~8u", binary:bin_to_list(Acc)) of
-        {error, _} -> throw({error, invalid_tar_checksum});
-        {ok, [Octal], []} -> Octal;
-        {ok, _, _} -> throw({error, invalid_tar_checksum})
-    end;
-do_parse_octal(<<$\s,Rest/binary>>, Acc) ->
+    Acc;
+do_parse_octal(<<$\s, Rest/binary>>, Acc) ->
     do_parse_octal(Rest, Acc);
 do_parse_octal(<<0, Rest/binary>>, Acc) ->
     do_parse_octal(Rest, Acc);
 do_parse_octal(<<C, Rest/binary>>, Acc) ->
-    do_parse_octal(Rest, <<Acc/binary, C>>).
+    Digit = C - $0,
+    if
+        0 =< Digit, Digit < 8 ->
+            do_parse_octal(Rest, Acc bsl 3 bor Digit);
+        true ->
+            throw({error, invalid_tar_checksum})
+    end.
 
 parse_string(Bin) when is_binary(Bin) ->
-    do_parse_string(Bin, <<>>).
-do_parse_string(<<>>, Acc) ->
-    case unicode:characters_to_list(Acc) of
+    N = strlen(Bin, 0),
+    <<Prefix:N/binary,_/binary>> = Bin,
+    case unicode:characters_to_list(Prefix) of
         Str when is_list(Str) ->
             Str;
         {incomplete, _Str, _Rest} ->
-            binary:bin_to_list(Acc);
+            binary_to_list(Bin);
         {error, _Str, _Rest} ->
             throw({error, {bad_header, invalid_string}})
-    end;
-do_parse_string(<<0, _/binary>>, Acc) ->
-    do_parse_string(<<>>, Acc);
-do_parse_string(<<C, Rest/binary>>, Acc) ->
-    do_parse_string(Rest, <<Acc/binary, C>>).
+    end.
+
+strlen(<<>>, N) -> N;
```

Comment:
probably worth renaming this function to be something like `length_until_zero` or something, I feel like `strlen` isn't really descriptive enough

Reply:
Yeah, naming is one of hardest problems. I tried a few names myself, until I settled on `strlen/1` because it is very similar to C's `strlen()`. I think it is clear enough from context.